qiskit/primitives/base_estimator.py

```diff
@@ -141,8 +148,11 @@ def __init__(
         Raises:
             QiskitError: For mismatch of circuits and parameters list.
         """
-        self._circuits = tuple(circuits)
-        self._observables = tuple(observables)
+        self._circuits = circuits
+        self._observables = observables
+        self._circuit_names = self._circuit_names
+        self._observable_ids = self._observable_ids
```

Comment:
These lines look redundant. Are they necessary?

Reply:
I think this is the only way to guarantee that they exist as class attributes. With only dynamic set, the python will not know if the attribute exists or not, so it must be presented in the constructor.

Reply:
How about `__slots__` to guarantee the attributes?
https://docs.python.org/3.9/reference/datamodel.html#object.__slots__

Reply:
Even if `__slots__` is added like `__slot__ = ("_circuits", "_observables", "_parameters", "_circuit_names", "_observable_ids")` in BaseEstimator, pylint raises no-member errors.
```
qiskit/primitives/base_estimator.py:291:45: E1101: Instance of 'BaseEstimator' has no '_circuit_names' member (no-member)
qiskit/primitives/base_estimator.py:303:47: E1101: Instance of 'BaseEstimator' has no '_observable_ids' member (no-member)
```